smd/main.go

```diff
@@ -77,26 +102,140 @@ func (o *options) resolve() (operation, error) {
 
 	switch {
 	case o.validatePath != "":
+		if o.merge || o.compare {
+			return nil, ErrTooManyOperations
+		}
 		return validation{base, o.validatePath}, nil
+	case o.merge:
+		if o.compare {
+			return nil, ErrTooManyOperations
+		}
+		if o.lhsPath == "" || o.rhsPath == "" {
+			return nil, ErrNeedTwoArgs
+		}
+		return merge{base, o.lhsPath, o.rhsPath}, nil
+	case o.compare:
+		if o.lhsPath == "" || o.rhsPath == "" {
+			return nil, ErrNeedTwoArgs
+		}
+		return compare{base, o.lhsPath, o.rhsPath}, nil
 	}
 	return nil, errors.New("no operation requested")
 }
 
+func (o *options) openOutput() (io.WriteCloser, error) {
+	if o.output == "-" {
+		return os.Stdout, nil
+	}
+	f, err := os.Create(o.output)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open %q for writing: %v", o.output, err)
+	}
+	return f, nil
+}
+
+func (b operationBase) parseFile(path string) (tv typed.TypedValue, err error) {
+	bytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return tv, fmt.Errorf("unable to read file %q: %v", path, err)
+	}
+	tv, err = b.parser.FromYAML(typed.YAMLObject(bytes), b.typeName)
+	if err != nil {
+		return tv, fmt.Errorf("unable to validate file %q:\n%v", path, err)
+	}
+	return tv, nil
+}
+
 type validation struct {
 	operationBase
 
 	fileToValidate string
 }
 
-func (v validation) execute() error {
-	b, err := ioutil.ReadFile(v.fileToValidate)
+func (v validation) execute(_ io.Writer) error {
+	_, err := v.parseFile(v.fileToValidate)
+	return err
+}
+
+type merge struct {
+	operationBase
+
+	lhs string
+	rhs string
+}
+
+func (m merge) execute(w io.Writer) error {
+	lhs, err := m.parseFile(m.lhs)
+	if err != nil {
+		return err
+	}
+	rhs, err := m.parseFile(m.rhs)
+	if err != nil {
+		return err
+	}
+
+	out, err := lhs.Merge(rhs)
+	if err != nil {
+		return err
+	}
+
+	yaml, err := out.AsValue().ToYAML()
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(yaml)
+
+	return err
+}
+
+type compare struct {
+	operationBase
+
+	lhs string
+	rhs string
+}
+
+func (c compare) execute(w io.Writer) error {
+	lhs, err := c.parseFile(c.lhs)
+	if err != nil {
+		return err
+	}
+	rhs, err := c.parseFile(c.rhs)
 	if err != nil {
-		return fmt.Errorf("unable to read file %q: %v", v.fileToValidate, err)
+		return err
 	}
-	_, err = v.parser.FromYAML(typed.YAMLObject(b), v.typeName)
+
+	got, err := lhs.Compare(rhs)
 	if err != nil {
-		return fmt.Errorf("unable to validate file %q:\n%v", v.fileToValidate, err)
+		return err
+	}
+
+	if got.Added.Empty() && got.Modified.Empty() && got.Removed.Empty() {
```

Comment:
unrelated but this is done a lot in tests, we might want to add a method like `Comparison.AreSame() bool` maybe

Reply:
done